File: web-api/providers/ucloud.mjs
import crypto from "node:crypto";
import { writeApiLog } from "../core/database.mjs";
import { decryptSecret } from "../core/crypto.mjs";
import { getAccountSecretRecord } from "../repositories/accounts.mjs";

function firstAddress(value) {
  if (Array.isArray(value)) return value.find((item) => item !== undefined && item !== null && String(item).trim() !== "") || "";
  return value === undefined || value === null ? "" : String(value);
}

export async function request(accountId, action, params = {}) {
  const row = getAccountSecretRecord(accountId);
  if (!row) throw new Error("云账号不存在");
  if (!row.enabled) throw new Error("云账号已停用");
  if (row.cloud_type !== "ucloud") throw new Error("当前账号不是 UCloud 账号");
  const query = { Action: action, PublicKey: row.access_key_id, ...params };
  const plain = Object.keys(query).sort().map((key) => `${key}${query[key]}`).join("") + decryptSecret(row.secret_ciphertext);
  query.Signature = crypto.createHash("sha1").update(plain).digest("base64");
  const response = await fetch(`https://api.ucloud.cn/?${new URLSearchParams(query)}`);
  const data = await response.json().catch(() => ({}));
  if (!response.ok || Number(data?.RetCode || 0) !== 0) { const message = data?.Message || data?.message || `UCloud ${response.status}`; writeApiLog(accountId, "api.ucloud.cn", action, params, data, "失败", message); throw new Error(message); }
  writeApiLog(accountId, "api.ucloud.cn", action, params, data, "成功");
  return data;
}

export async function pages(accountId, action, region, keys) {
  const items = [];
  for (let offset = 0; offset < 100_000; offset += 100) {
    const data = await request(accountId, action, { Region: region, Offset: offset, Limit: 100 });
    const page = keys.flatMap((key) => Array.isArray(data?.[key]) ? data[key] : []);
    items.push(...page);
    if (page.length < 100 || items.length >= Number(data?.TotalCount || data?.Total || Infinity)) return items;
  }
  throw new Error("分页超过 1000 页，已停止读取");
}

function instance(item, region) { return { ...item, InstanceId: item.UHostId, InstanceName: item.Name || item.UHostId, InstanceStatus: item.State, Status: item.State, PublicIpAddress: firstAddress(item.IPSet?.filter((ip) => ip.Type === "EIP") || item.IPSet), PrivateIpAddress: firstAddress(item.IPSet?.filter((ip) => ip.Type !== "EIP")), InstanceType: item.UHostType || item.CPU || "", VpcId: item.VPCId || "", _region_id: region }; }
function rds(item, region) { return { ...item, DBInstanceId: item.DBId, DBInstanceDescription: item.Name || item.DBId, DBInstanceStatus: item.State, DBInstanceClass: item.MemoryLimit || item.DBType || "", DBInstanceStorage: Number(item.DiskSpace || 0), ConnectionString: item.VirtualIP || "", Port: item.Port || "", Engine: item.DBType || "", EngineVersion: item.DBVersion || "", CreateTime: item.CreateTime || "", _region_id: region }; }
function redis(item, region) { return { ...item, InstanceId: item.GroupId, InstanceName: item.Name || item.GroupId, InstanceStatus: item.State, InstanceType: "Redis", InstanceClass: item.MemoryLimit || "", Capacity: Number(item.MemoryLimit || 0), ConnectionDomain: item.VirtualIP || item.VIP || "", Port: item.Port || "", EngineVersion: item.Version || "", NetworkType: item.VPCId || "", _region_id: region }; }
function bucket(item, region) { const name = item.BucketName || item.Name; return { ...item, Name: name, BucketName: name, Location: item.Region || region, CreationDate: item.CreateTime || "", StorageClass: item.StorageClass || "STANDARD", Acl: item.ACL || "private", ExtranetEndpoint: "-", IntranetEndpoint: "-", _region_id: item.Region || region }; }
function zone(item) { return { ...item, DomainName: item.DomainName || item.Domain, DomainStatus: item.Status || "ACTIVE", ZoneId: item.DomainId || item.DomainName, RecordCount: Number(item.RecordCount || 0), RegistrationDate: item.CreateTime || "", _region_id: "global", _ucloud_dns: true }; }

export async function resources(accountId, type, regions = ["cn-bj2"]) {
  const items = [];
  const errors = [];
  if (type === "domain") {
    try { items.push(...(await request(accountId, "DescribeUDNSDomain", { Offset: 0, Limit: 100 })).DomainSet || []); }
    catch (error) { errors.push(error.message); }
    return { resource_type: type, items: items.map(zone), errors, fetched_at: Date.now() };
  }
  const definitions = { ecs: ["DescribeUHostInstance", ["UHostSet"], instance], rds: ["DescribeUDBInstance", ["DataSet"], rds], redis: ["DescribeURedisGroup", ["DataSet"], redis], oss: ["DescribeUFileBucket", ["DataSet"], bucket] };
  const definition = definitions[type];
  if (!definition) return { resource_type: type, items, errors: [`UCloud 暂未接入 ${type} 资源`], fetched_at: Date.now() };
  const [action, keys, normalize] = definition;
  for (const region of regions) {
    try { items.push(...(await pages(accountId, action, region, keys)).map((item) => normalize(item, region))); }
    catch (error) { errors.push(`${region}: ${error.message}`); }
  }
  return { resource_type: type, items, errors, fetched_at: Date.now() };
}

export async function verify(accountId, regions = ["cn-bj2"]) {
  await request(accountId, "DescribeUHostInstance", { Region: regions[0], Offset: 0, Limit: 1 });
  return { provider: "ucloud", verified: true, region_count: regions.length, regions, default_region: regions[0] };
}

File: src-tauri/src/cloud/ucloud.rs
use crate::{account_cloud_type, account_credentials, array_at, configured_regions, rpc_encode, string_params, write_api_log, ResourceResponse};
use base64::{engine::general_purpose::STANDARD as B64, Engine as _};
use chrono::Utc;
use serde_json::{json, Value};
use sha1::{Digest, Sha1};
use std::collections::BTreeMap;

async fn request(id: i64, action: &str, mut params: BTreeMap<String, String>) -> Result<Value, String> {
    if account_cloud_type(id)? != "ucloud" { return Err("当前账号不是 UCloud 账号".into()); }
    let (access_key_id, access_key_secret) = account_credentials(id)?;
    params.insert("Action".into(), action.into());
    params.insert("PublicKey".into(), access_key_id.clone());
    let plain = params.iter().map(|(key, value)| format!("{key}{value}")).collect::<String>() + &access_key_secret;
    params.insert("Signature".into(), B64.encode(Sha1::digest(plain.as_bytes())));
    let query = params.iter().map(|(key, value)| format!("{}={}", rpc_encode(key), rpc_encode(value))).collect::<Vec<_>>().join("&");
    let response = reqwest::Client::new().get(format!("https://api.ucloud.cn/?{query}")).timeout(std::time::Duration::from_secs(30)).send().await.map_err(|error| format!("UCloud 请求失败: {error}"))?;
    let status = response.status();
    let data: Value = response.json().await.map_err(|error| format!("UCloud 返回解析失败: {error}"))?;
    if !status.is_success() || data.get("RetCode").and_then(Value::as_i64).unwrap_or(0) != 0 {
        let message = data.get("Message").or_else(|| data.get("message")).and_then(Value::as_str).unwrap_or("UCloud API 返回错误").to_string();
        write_api_log(&access_key_id, "api.ucloud.cn", action, &json!(params), Some(&data), "失败", Some(&message));
        return Err(message);
    }
    write_api_log(&access_key_id, "api.ucloud.cn", action, &json!(params), Some(&data), "成功", None);
    Ok(data)
}

async fn pages(id: i64, action: &str, region: &str, keys: &[&str]) -> Result<Vec<Value>, String> {
    let mut items = Vec::new();
    for offset in (0..100_000).step_by(100) {
        let data = request(id, action, string_params(&[("Region", region.into()), ("Offset", offset.to_string()), ("Limit", "100".into())])).await?;
        let page = keys.iter().flat_map(|key| array_at(&data, &[*key]).into_iter().cloned()).collect::<Vec<_>>();
        let count = page.len();
        items.extend(page);
        let total = data.get("TotalCount").or_else(|| data.get("Total")).and_then(Value::as_u64).unwrap_or(u64::MAX) as usize;
        if count < 100 || items.len() >= total { return Ok(items); }
    }
    Err("分页超过 1000 页，已停止读取".into())
}

fn first_ip(item: &Value, public: bool) -> Value {
    let Some(values) = item.get("IPSet").and_then(Value::as_array) else { return json!(""); };
    values.iter().find(|value| value.get("Type").and_then(Value::as_str).map(|kind| (kind == "EIP") == public).unwrap_or(!public)).and_then(|value| value.get("IP").or_else(|| value.get("Ip"))).cloned().unwrap_or(json!(""))
}

fn instance(item: &Value, region: &str) -> Value { json!({"InstanceId": item.get("UHostId"), "InstanceName": item.get("Name").or_else(|| item.get("UHostId")), "InstanceStatus": item.get("State"), "Status": item.get("State"), "PublicIpAddress": first_ip(item, true), "PrivateIpAddress": first_ip(item, false), "InstanceType": item.get("UHostType").or_else(|| item.get("CPU")), "VpcId": item.get("VPCId"), "_region_id": region, "_raw": item}) }
fn rds(item: &Value, region: &str) -> Value { json!({"DBInstanceId": item.get("DBId"), "DBInstanceDescription": item.get("Name").or_else(|| item.get("DBId")), "DBInstanceStatus": item.get("State"), "DBInstanceClass": item.get("MemoryLimit").or_else(|| item.get("DBType")), "DBInstanceStorage": item.get("DiskSpace").unwrap_or(&json!(0)), "ConnectionString": item.get("VirtualIP"), "Port": item.get("Port"), "Engine": item.get("DBType"), "EngineVersion": item.get("DBVersion"), "CreateTime": item.get("CreateTime"), "_region_id": region, "_raw": item}) }
fn redis(item: &Value, region: &str) -> Value { json!({"InstanceId": item.get("GroupId"), "InstanceName": item.get("Name").or_else(|| item.get("GroupId")), "InstanceStatus": item.get("State"), "InstanceType": "Redis", "InstanceClass": item.get("MemoryLimit"), "Capacity": item.get("MemoryLimit").unwrap_or(&json!(0)), "ConnectionDomain": item.get("VirtualIP").or_else(|| item.get("VIP")), "Port": item.get("Port"), "EngineVersion": item.get("Version"), "NetworkType": item.get("VPCId"), "_region_id": region, "_raw": item}) }
fn bucket(item: &Value, region: &str) -> Value { let name = item.get("BucketName").or_else(|| item.get("Name")); json!({"Name": name, "BucketName": name, "Location": item.get("Region").unwrap_or(&json!(region)), "CreationDate": item.get("CreateTime"), "StorageClass": item.get("StorageClass").unwrap_or(&json!("STANDARD")), "Acl": item.get("ACL").unwrap_or(&json!("private")), "ExtranetEndpoint": "-", "IntranetEndpoint": "-", "_region_id": item.get("Region").unwrap_or(&json!(region)), "_raw": item}) }
fn zone(item: &Value) -> Value { json!({"DomainName": item.get("DomainName").or_else(|| item.get("Domain")), "DomainStatus": item.get("Status").unwrap_or(&json!("ACTIVE")), "ZoneId": item.get("DomainId").or_else(|| item.get("DomainName")), "RecordCount": item.get("RecordCount").unwrap_or(&json!(0)), "RegistrationDate": item.get("CreateTime"), "_region_id": "global", "_ucloud_dns": true, "_raw": item}) }

pub(crate) async fn resource_items(id: i64, resource_type: &str) -> ResourceResponse {
    let now = Utc::now().timestamp_millis();
    let regions = match configured_regions(id, "cn-bj2") { Ok(value) => value, Err(error) => return ResourceResponse { resource_type: resource_type.into(), items: vec![], errors: vec![error], fetched_at: now } };
    let mut items = Vec::new();
    let mut errors = Vec::new();
    if resource_type == "domain" {
        match request(id, "DescribeUDNSDomain", string_params(&[("Offset", "0".into()), ("Limit", "100".into())])).await {
            Ok(data) => items.extend(array_at(&data, &["DomainSet"]).into_iter().map(zone)),
            Err(error) => errors.push(error),
        }
        return ResourceResponse { resource_type: resource_type.into(), items, errors, fetched_at: now };
    }
    let service = match resource_type { "ecs" => Some(("DescribeUHostInstance", vec!["UHostSet"])), "rds" => Some(("DescribeUDBInstance", vec!["DataSet"])), "redis" => Some(("DescribeURedisGroup", vec!["DataSet"])), "oss" => Some(("DescribeUFileBucket", vec!["DataSet"])), _ => None };
    let Some((action, keys)) = service else { return ResourceResponse { resource_type: resource_type.into(), items, errors: vec![format!("UCloud 暂未接入 {resource_type} 资源")], fetched_at: now }; };
    for region in regions {
        match pages(id, action, &region, &keys).await {
            Ok(values) => for item in values { items.push(match resource_type { "ecs" => instance(&item, &region), "rds" => rds(&item, &region), "redis" => redis(&item, &region), "oss" => bucket(&item, &region), _ => item }); },
            Err(error) => errors.push(format!("{region}: {error}")),
        }
    }
    ResourceResponse { resource_type: resource_type.into(), items, errors, fetched_at: now }
}

#[tauri::command]
pub(crate) async fn verify_ucloud_account(id: i64) -> Result<Value, String> {
    let regions = configured_regions(id, "cn-bj2")?;
    request(id, "DescribeUHostInstance", string_params(&[("Region", regions[0].clone()), ("Offset", "0".into()), ("Limit", "1".into())])).await?;
    Ok(json!({"provider":"ucloud","verified":true,"region_count":regions.len(),"regions":regions,"default_region":configured_regions(id, "cn-bj2")?[0]}))
}
